### PR TITLE
[WIP] Python API

### PR DESCRIPTION
**Pull request by bedapisl**
OUTDATED:
Here is a draft of Python API, any comments appreciated. The API does not have `MainLoop` class. Some of its functionality is IMHO not needed in the API (loop over epochs), some can be moved in the StreamWrapper (`buffer`, `fixed_epoch_size`, `fixed_batch_size`), some can be removed (`on_empty_batch`, `on_empty_stream`, `on_incorrect_config`) and some is in the `train` and `predict` functions (`on_unused_sources`).

Also I would like to change following:

- `eltf_freeze` creates new output directory and saves the `.pb` file and `config.yaml` there instead of changing config in already existing directory. I have two reasons for this:
 1. It is not possible to load the original model, when it was frozen (because its config was changed)
 2. When evaluating the frozen model, the outputs are in new directory. This means that the frozen model is in different directory than its outputs.

    Or we can move the  freezeing functionality to emloop, with syntax similar to `emloop eval`.
- rename `log_dir` to `output_dir` everywhere so that one thing has one name in all the code
- does any hook uses `self._main_loop`? I couldn't find one, so I think we can remove the `register_mainloop`.

### CODE REVIEW

File: emloop/api.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+def start(hooks):
+# calls before_training() for all hooks
+
+
+def train(stream: Iterable, model, hooks, on_unused_sources="warn"):
+# one epoch
+
+
+def predict(stream: Iterable, model, hooks, on_unused_sources="warn"):
```
Comment (petrbel) [inline emloop/api.py]:
```suggestion
def eval(stream: Iterable, model, hooks, on_unused_sources="warn"):
```

File: emloop/api.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+from .utils.config import load_yaml
+from .utils.reflection import parse_fully_qualified_name, create_object
+from .cli.util import find_config
+
+
+def load_yaml(path: str):
+# returns contents of YAML file as CommentedMap
+
+
+def create_output_dir(config: CommentedMap):
```
Comment (FloopCZ) [inline emloop/api.py]: What about an optional `pattern` parameter here so that the user is able to choose the destination and directory name? (see e.g., `template` parameter of `mktemp`: https://www.gnu.org/software/autogen/mktemp.html)

File: emloop/api.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+
+def create_dataset(config: CommentedMap, output_dir: Optional[str]=None):
+
+
+def create_hooks(config: CommentedMap, model=None, dataset=None, output_dir=None):
+# creates hooks and returns them as list
+
+
+def create_all(config: CommentedMap):
```
Comment (FloopCZ) [inline emloop/api.py]: I am not sure about the name of this function (the `all` part to be specific) but I cannot come up with anything better at the moment. Also, I don't like the quadruple as a return type and I would prefer a small class or a [named tuple](https://docs.python.org/3.8/library/collections.html#collections.namedtuple) instead.

File: emloop/api.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    hooks = create_hooks(config, model=model, dataset=dataset, output_dir=output_dir)
+
+    return output_dir, dataset, model, hooks
+
+
+def load_all(restore_path: str):
+# loads model, creates new output_dir, hooks and  dataset according to config in restore_path
+
+
+def start(hooks):
```
Comment (FloopCZ) [inline emloop/api.py]: I would rather have `start` and `finish` as a resource block (i.e., `with emloop.session(...) as sess`).

**Review by FloopCZ — COMMENTED** [review]
Thank you @bedapisl for working on this.

**Comment by bedapisl** [comment]
TODO: Add example of usage.

File: emloop/cli/common.py [hunk 5]
```diff
@@ -213,6 +213,26 @@ def create_hooks(config: dict, model: AbstractModel,
...
 
+def create_main_loop(config: dict, model: Optional[AbstractModel]=None, dataset: Optional[AbstractDataset]=None,
+                     hooks: Optional[List[AbstractHook]]=None, eval: Optional[str]=None):
+    """
+    Create a main loop object.
+
+    :param config: config dict with main loop config
+    :param dataset: dataset object implementing the :py:class:`emloop.datasets.AbstractDataset` concept
+    :param model: model object implementing the :py:class:`emloop.models.AbstractModel` concept
+    :param hooks: list of hooks implementing the :py:class:`emloop.models.AbstractHook` concept
```
Comment (FloopCZ) [inline emloop/cli/common.py:224]: `eval` is undocumented

File: emloop/cli/common.py [hunk 6]
```diff
@@ -213,6 +213,26 @@ def create_hooks(config: dict, model: AbstractModel,
...
+
+    :return: main loop object
+    """
+
+    logging.info('Creating main loop')
+    kwargs = config['main_loop'] if 'main_loop' in config else {}
```
Comment (FloopCZ) [inline emloop/cli/common.py]: You can do `config.get('main_loop', {})`.

File: emloop/cli/common.py [hunk 7]
```diff
@@ -213,6 +213,26 @@ def create_hooks(config: dict, model: AbstractModel,
...
+    if eval is not None:
+        kwargs['extra_streams'] = []
```
Comment (FloopCZ) [inline emloop/cli/common.py]: This will mutate the `extra_streams` item in the original dict, have you considered making a deep copy instead?

File: emloop/cli/common.py [hunk 8]
```diff
@@ -213,6 +213,26 @@ def create_hooks(config: dict, model: AbstractModel,
     return hooks
 
 
+def create_main_loop(config: dict, model: Optional[AbstractModel]=None, dataset: Optional[AbstractDataset]=None,
+                     hooks: Optional[List[AbstractHook]]=None, eval: Optional[str]=None):
```
Comment (FloopCZ) [inline emloop/cli/common.py]: I don't really like the variable named `eval`, since it is a built-in function.

File: emloop/main_loop.py [hunk 9]
```diff
@@ -198,22 +216,27 @@ def _run_epoch(self, stream: StreamWrapper, train: bool) -> None:
...
 
-        :param stream: stream to be evaluated
-        :param stream_name: stream name
+        :param stream: stream to be evaluated (string (e.g. 'train') or StreamWrapper)
         """
-        self._run_epoch(stream=stream, train=False)
+         if isinstance(stream, str):
+            stream = self.get_stream(stream)
+
+       self._run_epoch(stream=stream, train=False)
```
Comment (FloopCZ) [inline emloop/main_loop.py]: misindented

File: emloop/main_loop.py [hunk 10]
```diff
@@ -198,22 +216,27 @@ def _run_epoch(self, stream: StreamWrapper, train: bool) -> None:
...
+    def train_by_stream(self, stream: Union[StreamWrapper, str]) -> None:
         """
         Train the model with the given stream.
 
-        :param stream: stream to train with
+        :param stream: stream to train with (string (e.g. 'train') or StreamWrapper)
         """
+        if isinstance(stream, str):
+            stream = self.get_stream(stream)
+
```
Comment (FloopCZ) [inline emloop/main_loop.py:227]: You can delete this newline.

File: emloop/main_loop.py [hunk 11]
```diff
@@ -198,22 +216,27 @@ def _run_epoch(self, stream: StreamWrapper, train: bool) -> None:
...
         Evaluate the given stream.
 
-        :param stream: stream to be evaluated
-        :param stream_name: stream name
+        :param stream: stream to be evaluated (string (e.g. 'train') or StreamWrapper)
         """
-        self._run_epoch(stream=stream, train=False)
+         if isinstance(stream, str):
+            stream = self.get_stream(stream)
+
```
Comment (FloopCZ) [inline emloop/main_loop.py:239]: ditto

File: docs/advanced/python.rst [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+
+.. code-block:: python
+    create_output_dir
+    create_dataset
+    create_model
+    create_hooks
+    create_main_loop
+
+All these functions require configuration dictionary, which you can get from standard YAML config file using function `load_yaml` defined in `emloop/utils/yaml.py`.
+Then you can use methods `main_loop.train_by_stream` and `main_loop.evaluate_stream` for training and evaluating one epoch.
```
Comment (FloopCZ) [inline docs/advanced/python.rst]: Thank you for mentioning this in the documentation. However, this will surely need more love. There has to be an example, the functions in the list need to link to the API docs and the whole Python API should be directly under `emloop` package or under a dedicated subbackage for simpler usage.

**Comment by FloopCZ** [comment]
Thank you @bedapisl, one thing I do not like much is function name `train_by_stream`, what is your opinion? I am inclined to just `train` and `evaluate` instead of `train_by_stream` and `evaluate_stream`. The `stream` is a parameter of the function, so the semantics is pretty clear IMHO.

File: emloop/__init__.py [hunk 13]
```diff
@@ -6,6 +6,7 @@
 from .models import AbstractModel
 from .types import Batch, Stream, EpochData, TimeProfile
 
+from . import api
```
Comment (FloopCZ) [inline emloop/__init__.py]: Please add this module to the list of installed packages: https://github.com/iterait/emloop/blob/api/setup.py#L26 as well.

File: emloop/cli/common.py [hunk 14]
```diff
@@ -213,7 +213,28 @@ def create_hooks(config: dict, model: AbstractModel,
     return hooks
 
 
-def run(config: dict, output_root: str, restore_from: str=None, eval: Optional[str]=None) -> None:
+def create_main_loop(config: dict, model: Optional[AbstractModel]=None, dataset: Optional[AbstractDataset]=None,
+                     hooks: Optional[List[AbstractHook]]=None, eval_stream: Optional[str]=None):
```
Comment (FloopCZ) [inline emloop/cli/common.py]: ~~Wouldn't it be better to pass `extra_streams` to be consistent with the corresponding config option?~~
Edit: As discussed in person, the `extra_streams` and `eval_stream` behavior is kind of ugly at the moment. Please @bedapisl consider whether this could be improved.

The documentation for this parameter should also say what is the default behavior.

File: emloop/api.py (hunk 2 above)
Comment (blazekadam) [reply emloop/api.py]: :+1: the default pattern is imo quite long anyways

File: emloop/__init__.py (hunk 13 above)
Comment (blazekadam) [inline emloop/__init__.py]: I would like to have access to the `api` methods from the uppermost module.

Please consider
```python
from .api import *
```
and expose the methods in `__all__`.

File: emloop/cli/common.py [hunk 15]
```diff
@@ -91,10 +91,10 @@ def create_dataset(config: dict, output_dir: Optional[str]=None) -> AbstractData
     return dataset
 
 
-def create_model(config: dict, output_dir: Optional[str], dataset: AbstractDataset,
+def create_model(config: dict, output_dir: Optional[str]=None, dataset: Optional[AbstractDataset]=None,
                  restore_from: Optional[str]=None) -> AbstractModel:
     """
-    Create a model object either from scratch of from the checkpoint in ``resume_dir``.
+    Create a model object either from scratch or from the checkpoint in ``resume_dir``.
```
Comment (blazekadam) [inline emloop/cli/common.py]: in ``restore_from``.

File: emloop/cli/common.py [hunk 16]
```diff
@@ -279,19 +297,21 @@ def run(config: dict, output_root: str, restore_from: str=None, eval: Optional[s
         fallback('Creating hooks failed', ex)
 
     try:
-        logging.info('Creating main loop')
-        kwargs = config['main_loop'] if 'main_loop' in config else {}
-        if eval is not None:
-            kwargs['extra_streams'] = []
-        main_loop = MainLoop(model=model, dataset=dataset, hooks=hooks, **kwargs)
+        if eval_stream is not None:
+            if "main_loop" not in config_copy:
```
Comment (blazekadam) [inline emloop/cli/common.py]:
```suggestion
            if 'main_loop' not in config_copy:
```

File: emloop/cli/common.py [hunk 17]
```diff
@@ -279,19 +297,21 @@ def run(config: dict, output_root: str, restore_from: str=None, eval: Optional[s
...
+                config_copy["main_loop"] = {}
```
Comment (blazekadam) [inline emloop/cli/common.py]:
```suggestion
                config_copy['main_loop'] = {}
```

File: emloop/cli/common.py [hunk 18]
```diff
@@ -279,19 +297,21 @@ def run(config: dict, output_root: str, restore_from: str=None, eval: Optional[s
...
+
+            config_copy["main_loop"]["extra_streams"] = []
```
Comment (blazekadam) [inline emloop/cli/common.py]:
```suggestion
            config_copy['main_loop']['extra_streams'] = []
```

File: emloop/cli/common.py [hunk 19]
```diff
@@ -91,10 +91,10 @@ def create_dataset(config: dict, output_dir: Optional[str]=None) -> AbstractData
     return dataset
 
 
-def create_model(config: dict, output_dir: Optional[str], dataset: AbstractDataset,
+def create_model(config: dict, output_dir: Optional[str]=None, dataset: Optional[AbstractDataset]=None,
```
Comment (blazekadam) [inline emloop/cli/common.py:94]: I would love having all the `create_*` methods accepting both config filename (str) and config dict (otherwise/dict).

File: emloop/cli/common.py [hunk 20]
```diff
@@ -244,12 +262,12 @@ def run(config: dict, output_root: str, restore_from: str=None, eval: Optional[s
...
     :param output_root: dir under which output_dir shall be created
     :param restore_from: from whence the model should be restored (backend-specific information)
-    :param eval: optional name of the stream to be evaluated
+    :param eval_stream: Optional name of the stream to be evaluated. If set, extra_streams of main_loop are ignored.
     """
 
     output_dir = dataset = model = hooks = main_loop = None
 
-    config_copy = deepcopy(config)
+    config_copy = copy.deepcopy(config)
```
Comment (blazekadam) [inline emloop/cli/common.py:270]: I wonder why did you change this (from `deepcopy` to `copy.deepcopy`)

File: emloop/cli/common.py [hunk 21]
```diff
@@ -213,7 +213,25 @@ def create_hooks(config: dict, model: AbstractModel,
...
+
+    :return: main loop object
+    """
+
+    logging.info('Creating main loop')
+    main_loop_kwargs = copy.deepcopy(config.get('main_loop', {}))
+    return MainLoop(model=model, dataset=dataset, hooks=hooks, **main_loop_kwargs)
+
+
+def run(config: dict, output_root: str, restore_from: str=None, eval_stream: Optional[str]=None) -> None:
```
Comment (blazekadam) [inline emloop/cli/common.py:234]: Do we need to modify (and expose) this method? Basically it does something similar to *create_all* and then calls a single main loop's method. I would rather force the user to create the main loop and call the method for himself instead of *hacking* the config...

File: emloop/main_loop.py [hunk 22]
```diff
@@ -98,6 +98,24 @@ def __init__(self,   # pylint: disable=too-many-arguments
...
 
+    def __enter__(self):
+        """Calls before_training() for all hooks."""
+        self.hooks_before_training()
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Calls after_training() for all hooks."""
+        self.hooks_after_training()
+
+    def hooks_before_training(self):
```
Comment (blazekadam) [inline emloop/main_loop.py]: I would name this `trigger_before_training` (similarly for the following method).

File: emloop/main_loop.py [hunk 23]
```diff
@@ -250,8 +275,7 @@ def _run_zeroth_epoch(self, streams: Iterable[str]) -> None:
 
         self._epoch_profile.clear()
         for stream_name in streams:
-            with self.get_stream(stream_name) as stream:
-                self.evaluate_stream(stream)
+            self.evaluate(stream_name)
```
Comment (blazekadam) [inline emloop/main_loop.py:278]: whoa, now you do not call `__enter__` and `__exit__` on the `StreamWrapper`, this may break the buffering I believe.

File: emloop/main_loop.py [hunk 24]
```diff
@@ -314,12 +335,10 @@ def training():
                 self._epoch_profile.clear()
                 epoch_data = self._create_epoch_data()
 
-                with self.get_stream(self._train_stream_name) as stream:
-                    self.train_by_stream(stream)
+                self.train(self._train_stream_name)
```
Comment (blazekadam) [inline emloop/main_loop.py:338]: ditto

File: emloop/main_loop.py [hunk 25]
```diff
@@ -314,12 +335,10 @@ def training():
...
 
                 for stream_name in self._extra_streams:
-                    with self.get_stream(stream_name) as stream:
-                        self.evaluate_stream(stream)
+                    self.evaluate(stream_name)
```
Comment (blazekadam) [inline emloop/main_loop.py:341]: ditto

**Review by blazekadam — CHANGES_REQUESTED** [review]

File: emloop/cli/common.py (hunk 19 above)
Comment (bedapisl) [reply emloop/cli/common.py:94]: I don't agree with this. I think there should be only one way how to do something with an API not two ways. And part of the API is function `load_yaml` which allows you to convert the filename to dict.

File: emloop/cli/common.py (hunk 21 above)
Comment (bedapisl) [reply emloop/cli/common.py:234]: I did not modify this function, I only renamed one of its arguments as Filip told me to do in his review. Also this function is not exposed in the API.

File: emloop/main_loop.py (hunk 23 above)
Comment (bedapisl) [reply emloop/main_loop.py:278]: I do, look at lines: https://github.com/iterait/emloop/pull/58/files/ce25ab0a86dbcf34cc7b44f936a3fd03f5fa9caa#diff-83cf5394877642e7670e44b807f8858bR228 and 
https://github.com/iterait/emloop/pull/58/files/ce25ab0a86dbcf34cc7b44f936a3fd03f5fa9caa#diff-83cf5394877642e7670e44b807f8858bR240.

File: emloop/cli/common.py (hunk 20 above)
Comment (FloopCZ) [reply emloop/cli/common.py:270]: I actually like this one. :wink: 

> Namespaces are one honking great idea -- let's do more of those!
>                      `import this`

File: emloop/main_loop.py (hunk 23 above)
Comment (blazekadam) [reply emloop/main_loop.py:278]: my bad

File: emloop/cli/common.py (hunk 20 above)
Comment (blazekadam) [reply emloop/cli/common.py:270]: I sort of like it as well yet I still wonder you see...

File: emloop/cli/common.py (hunk 21 above)
Comment (blazekadam) [reply emloop/cli/common.py:234]: Well ,there are some changes as you create the main loop via `create_main_loop`, perhaps the old direct way was better. I do not like the conditional config dict updates.

Comment (FloopCZ) [reply emloop/cli/common.py:234]: Yeah, we discussed the conditional update with beda and we dont like it either.

Comment (bedapisl) [reply emloop/cli/common.py:234]: I think the current version can be merged. If you don't like it for some reason, please suggest an alternative.

Comment (FloopCZ) [reply emloop/cli/common.py:234]: Nah, we need to change this. Will think about it tommorrow.

**Comment by bedapisl** [comment]
Closing in favor of #64.
